risc0/zkvm/src/host/server/exec/syscall/prove_zkr.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Manages system calls for accelerators and other proof composition
+
+use anyhow::Result;
+use risc0_circuit_rv32im::prove::emu::addr::ByteAddr;
+use risc0_zkvm_platform::{
+    syscall::reg_abi::{REG_A3, REG_A4, REG_A5, REG_A6, REG_A7},
+    WORD_SIZE,
+};
+
+use crate::{
+    host::client::env::ProveZkrRequest, recursion::prove::get_registered_zkr, Assumption,
+    AssumptionReceipt,
+};
+
+use super::{Syscall, SyscallContext};
+
+#[derive(Clone)]
+pub(crate) struct SysProveZkr;
+
+impl Syscall for SysProveZkr {
+    fn syscall(
+        &mut self,
+        _syscall: &str,
+        ctx: &mut dyn SyscallContext,
+        _to_guest: &mut [u32],
+    ) -> Result<(u32, u32)> {
+        let claim_digest = ctx.load_digest_from_register(REG_A3)?;
+        let control_id = ctx.load_digest_from_register(REG_A4)?;
+        let control_root = ctx.load_digest_from_register(REG_A5)?;
+        let input_ptr = ByteAddr(ctx.load_register(REG_A6));
+        let input_len = ctx.load_register(REG_A7);
+        let input: Vec<u8> = ctx.load_region(input_ptr, input_len * WORD_SIZE as u32)?;
+
+        let proof_request = ProveZkrRequest {
+            claim_digest,
+            control_id,
+            input,
+        };
+
+        if let Some(coprocessor) = &ctx.syscall_table().coprocessor {
+            coprocessor.borrow_mut().prove_zkr(proof_request)?;
+        } else {
+            get_registered_zkr(&control_id)?;
+            ctx.syscall_table()
+                .pending_zkrs
+                .borrow_mut()
+                .push(proof_request);
+        }
```

Comment:
With this implementation, there is a bit of a dilemma for developers that want to add a new coprocessor to their proving context in that they need to provide a single handler for _all_ coprocessor requests. In the case of client-server with IPC, and the ZKR registry being a process-level mapping, this may be an issue. E.g. if someone is using `r0vm` through `LocalProver`, it might have the the RSA and Keccak coprocessors installed in the `r0vm` process' registry. If the client wants to add ED25519 verify (possibly a ZKR they wrote) the callback mechanism provides a way to do this. However, adding their coprocessor effectively "uninstalls" RSA and Keccak because all coprocessor requests will now be forwarded to the client. If they want to use those, they need to go install those as well.

Since we are focused on integration with Bonsai and other prover services here, that will want to take over ZKR proving with custom code, maybe this fits the needs of the use cases at hand. Another option here would be to have a mapping of `control_root` to coprocessor handler, such that each new coprocessor could indicate which requests it's taking ownership of. A default handler could be installed, which would take all requests not taken by a more specific handler.

Reply:
Let's go with this for now and see how well it integrates with Bonsai. Then I think we can come back and see if we even want to add the ability for users to add their own callback handlers. I actually started with this approach (but the key was `control_id` rather than `control_root`) and the part that was annoying was how to register the handlers to accept a wildcard (i.e. all requests like it is now).